TESTING.md

```diff
@@ -36,20 +36,156 @@ vcd:
 
 Users must specify their username, password, api_endpoint, vcd and org for any tests to run. Otherwise all tests get aborted. For more comprehensive testing the catalog, catalogitem, storageprofile, network, edgegateway, ip field can be set using the format above. For comprehensive testing just replace each field with your vcd information. 
 
-# Running Tests
+## Running Tests
 Once you have a config file setup, you can run tests with either the makefile or with go itself.
 
 To run tests with go use these commands:
-```
+
+```bash
 cd govcd
 go test -check.v .
 ```
 
 To run tests with the makefile:
-```
+
+```bash
 make test
 ```
 
+## How to write a test
+
+go-vcloud-director tests are written using [check.v1](https://labix.org/gocheck), an auxiliary libarry for tests that provides several methods to help developers write comprehensive tests.
+
+
+### Imports
+
+The tests for `govcd` package must belong to the same package, so no need to import govcd.
+The mandatory import is
+
+```go
+    checks "gopkg.in/check.v1"
+```
+
+We refer to this package as `checks`. There are examples online using a dot (".") instead of an explicit name, but please don't do it, as it pollutes the name space and makes code readability harder.
+
+### Test function header
+
+Every function that tests something related to govcd should have this header:
+
+```go
+func (vcd *TestVCD) Test_SomeFunctionality(check *checks.C) {
+...
+}
+```
+
+The `vcd` variable is our handler of the govcd functions. Its type is declared within `./govcd/api_vcd_test.go` and contains accessors to the main structures needed for running tests:
+
+```go
+type TestVCD struct {
+	client *VCDClient
+	org    Org
+	vdc    Vdc
+	vapp   VApp
+	config TestConfig
+}
+```
+This entity is initialized when the test starts. You can assume that the variable is ready for usage. If something happens during initialization, it should fail before it comes to your function.
+
+The `check` variable is our interface with the `check.v1` package. We can do several things with it, such as skipping a test, probing a condition, declare success or failure, and more.
+
+
+### Basic test function organization.
+
+Within the testing function, you should perform four actions:
+
+1. Run all operations that are needed for the test, such as finding a particular organization, or vDC, or vApp, deploying a VM, etc.
+2. Run the operation being tested (such as deploy a VM, retrieve a VM or vApp, etc.)
+3. Run the tests, which usually means using `check.Assert` or `check.Check` to compare known data with what was found during the test execution
+4. Optionally clean up all entities and data that may affect other test runs (e.g. delete something just deployed)
```

Comment:
This shouldn't be optional. I think the wording is a little confusing. It should state that if creating entities and data please clean it up as tests should be idempotent. If entities are not cleaned then it might ruin the idempotent behavior or the usefulness of the test. Also tests should not create any lasting entities on an user's vcd.

Reply:
Thanks @auppunda . 
In fact, the wording is misleading, and it may look like the cleanup can be avoided. What I need to say instead is that the cleanup is mandatory when the test creates something new, while it is not needed when the test is read-only. 
I will make it more clear. Perhaps I will introduce a new heading for this purpose.